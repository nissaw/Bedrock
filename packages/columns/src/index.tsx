import styled from "styled-components";
import PropTypes from "prop-types";
import {
  spacing as defaultSpacings,
  SpacingTypes,
  mergeSpacings,
} from "@bedrock-layout/spacing-constants";

export interface ColumnsProps {
  gutter?: SpacingTypes;
  columns?: number;
  dense?: boolean;
}

const Columns = styled.div<ColumnsProps>`
  box-sizing: border-box;

  --gutter: ${({ gutter, theme: { spacing = {} } }) =>
    gutter && mergeSpacings(spacing)[gutter]
      ? mergeSpacings(spacing)[gutter]
      : mergeSpacings(spacing).lg};
  --columns: ${({ columns = 1 }) => (columns > 0 ? columns : 1)};

  display: grid;
  grid-template-columns: repeat(var(--columns), 1fr);
  grid-gap: var(--gutter);
  grid-auto-flow: row ${(props) => props.dense === true && "dense"};

  @supports not (grid-gap: var(--gutter)) {
    display: flex;
    flex-flow: column;

    > * + * {
      margin-top: ${({ gutter, theme: { spacing = {} } }) =>
        gutter && mergeSpacings(spacing)[gutter]
          ? mergeSpacings(spacing)[gutter]
          : mergeSpacings(spacing).lg};
    }
  }
`;

Columns.displayName = "Columns";

Columns.propTypes = {
  gutter: PropTypes.oneOf<SpacingTypes>(
    Object.keys(defaultSpacings) as SpacingTypes[]
  ),
  columns: PropTypes.number,
  dense: PropTypes.bool,
};

Columns.defaultProps = {
  gutter: "lg",
  columns: 1,
  dense: false,
};

export interface ColumnProps {
  span?: number;
}

type SafeSpan = (span: unknown) => number;
const safeSpan: SafeSpan = (span) => {
  return typeof span === "number" ? span : 1;
};

export const Column = styled.div.attrs((props) => ({
  /* eslint-disable */
  ["data-bedrock-layout-column"]: "",
}))<ColumnProps>`
  grid-column: span ${({ span = 1 }) => Math.max(safeSpan(span), 1)} / auto;
`;

Column.propTypes = {
  span: PropTypes.number,
};

Column.defaultProps = {
  span: 1,
};

if (typeof exports !== "undefined") {
  Object.defineProperty(exports, "__esModule", {
    value: true,
  });
}
export default Columns;
