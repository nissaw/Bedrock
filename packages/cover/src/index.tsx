import {
  SpacingTypes,
  spacing as defaultSpacings,
  mergeSpacings,
} from "@bedrock-layout/spacing-constants";
import { forwardRefWithAs } from "@bedrock-layout/type-utils";
import PropTypes from "prop-types";
import React, { Children } from "react";
import styled from "styled-components";

const VerticallyCentered = styled.div`
  margin-top: auto;
  margin-bottom: auto;
  margin-block-start: auto;
  margin-block-end: auto;
`;

const Top = styled.div`
  margin-top: 0;
  margin-block-start: 0;
  margin-bottom: var(--gutter);
  margin-block-end: var(--gutter);
`;

const Bottom = styled.div`
  margin-top: var(--gutter);
  margin-block-start: var(--gutter);
  margin-bottom: 0;
  margin-block-end: 0;
`;

interface CoverWrapperProps {
  gutter?: SpacingTypes;
  minHeight?: string;
}

const CoverWrapper = styled.div<CoverWrapperProps>`
  --gutter: ${({ gutter, theme: { spacing = {} } }) =>
    gutter && mergeSpacings(spacing)[gutter]
      ? mergeSpacings(spacing)[gutter]
      : mergeSpacings(spacing).lg};

  display: flex;
  flex-direction: column;
  min-height: ${(props) => props.minHeight || "100vh"};
`;

export interface CoverProps extends CoverWrapperProps {
  top?: React.ReactNode;
  bottom?: React.ReactNode;
}

const Cover = forwardRefWithAs<CoverProps, "div">(
  ({ children, top, bottom, as, ...props }, ref) => {
    return (
      <CoverWrapper as={as} ref={ref} {...props}>
        {top && <Top>{Children.only(top)}</Top>}
        <VerticallyCentered>{Children.only(children)}</VerticallyCentered>
        {bottom && <Bottom>{Children.only(bottom)}</Bottom>}
      </CoverWrapper>
    );
  }
);

Cover.displayName = "Cover";

Cover.propTypes = {
  gutter: PropTypes.oneOf<SpacingTypes>(
    Object.keys(defaultSpacings) as SpacingTypes[]
  ),
  minHeight: PropTypes.string,
  top: PropTypes.element,
};

export default Cover;
