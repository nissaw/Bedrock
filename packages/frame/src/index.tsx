import styled from "styled-components";
import PropTypes from "prop-types";

export interface FrameProps {
  ratio: [number, number];
  position?: string;
}

const Frame = styled.div<FrameProps>`
  --d: ${(props) =>
    props.ratio && props.ratio[0] && Number.isInteger(props.ratio[0])
      ? props.ratio[0]
      : 1};
  --n: ${(props) =>
    props.ratio && props.ratio[1] && Number.isInteger(props.ratio[1])
      ? props.ratio[1]
      : 1};
  box-sizing: border-box;
  display: block;
  width: 100%;
  padding-bottom: calc(var(--n) / var(--d) * 100%);
  position: relative;

  > * {
    overflow: hidden;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  > img,
  > video {
    width: 100%;
    height: 100%;
    object-fit: cover;
    object-position: ${(props) =>
      typeof props.position === "string" ? props.position : "50%"};
  }
`;

Frame.displayName = "Frame";

type TwoNumbers = (props: FrameProps, propName: string) => Error | undefined;
const twoNumbers: TwoNumbers = ({ ratio }, propName) => {
  if (typeof ratio === "undefined") return new Error(`${propName} is required`);

  if (
    !Array.isArray(ratio) ||
    ratio.length !== 2 ||
    !ratio.every(Number.isInteger)
  ) {
    console.error(`${propName} needs to be an array of two numbers`);
  }
  return undefined;
};

Frame.propTypes = {
  //It's valid propType but type of Validator<[number, number]> makes no sense
  ratio: (twoNumbers as unknown) as React.Validator<[number, number]>,
  position: PropTypes.string,
};

export default Frame;
