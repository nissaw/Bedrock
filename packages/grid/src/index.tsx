import {
  SpacingTypes,
  spacing as defaultSpacings,
  mergeBreakpoints,
  mergeSpacings,
} from "@bedrock-layout/spacing-constants";
import PropTypes from "prop-types";
import styled from "styled-components";

export interface GridProps {
  gutter?: SpacingTypes;
  minItemWidth?: number;
}

const Grid = styled.div<GridProps>`
  box-sizing: border-box;

  --gutter: ${({ gutter, theme: { spacing = {} } }) =>
    gutter && mergeSpacings(spacing)[gutter]
      ? mergeSpacings(spacing)[gutter]
      : mergeSpacings(spacing).lg};

  --minItemWidth: ${(props) =>
    typeof props.minItemWidth === "number"
      ? `${props.minItemWidth}px`
      : mergeBreakpoints(props.theme.breakPoints).smallOnly + "px"};

  display: grid;
  gap: var(--gutter);

  grid-template-columns: repeat(
    auto-fit,
    minmax(min(var(--minItemWidth), 100%), 1fr)
  );
`;

Grid.displayName = "Grid";

Grid.propTypes = {
  gutter: PropTypes.oneOf<SpacingTypes>(
    Object.keys(defaultSpacings) as SpacingTypes[]
  ),
  minItemWidth: PropTypes.number,
};

Grid.defaultProps = {
  gutter: "lg",
};

export default Grid;
