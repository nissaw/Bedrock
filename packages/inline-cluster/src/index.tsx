import {
  SpacingTypes,
  spacing as defaultSpacings,
  mergeSpacings,
} from "@bedrock-layout/spacing-constants";
import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";

type JustifyAlignOptions = "start" | "center" | "end";

type JustifyAlignMap = { [key in JustifyAlignOptions]: string };
const justifyAlignMap: JustifyAlignMap = {
  start: "flex-start",
  end: "flex-end",
  center: "center",
};
export interface InlineClusterProps {
  justify?: JustifyAlignOptions;
  align?: JustifyAlignOptions;
  gutter?: SpacingTypes;
}

const InlineCluster = styled.div.attrs(({ children }) => {
  return {
    children: React.Children.map(children, (child) => <div>{child}</div>),
  };
})<InlineClusterProps>`
  box-sizing: border-box;

  --gutter: ${({ gutter, theme: { spacing = {} } }) =>
    gutter && mergeSpacings(spacing)[gutter]
      ? mergeSpacings(spacing)[gutter]
      : mergeSpacings(spacing).lg};

  display: inline-flex;
  flex-wrap: wrap;
  inline-size: 100%;

  justify-content: ${(props) =>
    typeof props.justify !== "undefined" && justifyAlignMap[props.justify]
      ? justifyAlignMap[props.justify]
      : justifyAlignMap.start};

  align-items: ${(props) =>
    typeof props.align !== "undefined" && justifyAlignMap[props.align]
      ? justifyAlignMap[props.align]
      : justifyAlignMap.start};

  margin: calc(var(--gutter) / 2 * -1);

  & > * {
    margin: calc(var(--gutter) / 2);
  }
`;

InlineCluster.displayName = "InlineCluster";

InlineCluster.propTypes = {
  gutter: PropTypes.oneOf<SpacingTypes>(
    Object.keys(defaultSpacings) as SpacingTypes[]
  ),
  justify: PropTypes.oneOf<JustifyAlignOptions>(["start", "center", "end"]),
  align: PropTypes.oneOf<JustifyAlignOptions>(["start", "center", "end"]),
};

export default InlineCluster;
