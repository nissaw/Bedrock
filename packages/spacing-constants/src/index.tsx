const none = "0px";
const xs = "0.125rem";
const sm = "0.25rem";
const md = "0.5rem";
const mdLg = "0.75rem";
const lg = "1rem";
const lgXl = "1.5rem";
const xl = "2rem";
const xlXXl = "3rem";
const xxl = "4rem";

export interface Spacing {
  none: string;
  xs: string;
  sm: string;
  md: string;
  mdLg: string;
  lg: string;
  lgXl: string;
  xl: string;
  xlXXl: string;
  xxl: string;
}

export type SpacingTypes = keyof Spacing;

export const spacing: Spacing = {
  none,
  xs,
  sm,
  md,
  mdLg,
  lg,
  lgXl,
  xl,
  xlXXl,
  xxl,
};

type MergeSpacings = (spacing: object) => Spacing;
export const mergeSpacings: MergeSpacings = (newSpacings) => ({
  ...spacing,
  ...newSpacings,
});

type NumberTuple = [number, number];
/* Based on Foundations Break Points */
const smallOnly = 639;
const mediumOnly: NumberTuple = [640, 1023];
const largeOnly: NumberTuple = [1024, 1199];
const xlargeOnly: NumberTuple = [1200, 1439];
const medium = 1023;
const large = 1199;
const xlarge = 1439;
const xxlarge = 1920;

export interface BreakPoints {
  smallOnly: number;
  mediumOnly: NumberTuple;
  largeOnly: NumberTuple;
  xlargeOnly: NumberTuple;
  medium: number;
  large: number;
  xlarge: number;
  xxlarge: number;
}

export type BreakPointTypes = keyof BreakPoints;

export const breakPoints: BreakPoints = {
  smallOnly,
  mediumOnly,
  largeOnly,
  xlargeOnly,
  medium,
  large,
  xlarge,
  xxlarge,
};

type MergeBreakpoints = (breakPoints: object) => BreakPoints;
export const mergeBreakpoints: MergeBreakpoints = (newBreakPoints = {}) => ({
  ...breakPoints,
  ...newBreakPoints,
});
