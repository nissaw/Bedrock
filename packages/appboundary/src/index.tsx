import Center from "@bedrock-layout/center";
import { mergeBreakpoints } from "@bedrock-layout/spacing-constants";
import PropTypes from "prop-types";
import React from "react";
import styled, { ThemeContext } from "styled-components";

export interface AppBoundaryProps {
  children: React.ReactNode;
}

const Outer = styled.div`
  padding: 0;
  max-width: 100%;
  overflow: hidden;
  height: 100%;
`;

const safeTheme = { breakPoints: {} };

const AppBoundary = ({ children }: AppBoundaryProps) => {
  const { breakPoints = {} } = React.useContext(ThemeContext) || safeTheme;
  const maxWidth = mergeBreakpoints(breakPoints).xxlarge;
  return (
    <Outer>
      <Center maxWidth={maxWidth}>{children}</Center>
    </Outer>
  );
};

AppBoundary.displayName = "AppBoundary";

AppBoundary.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]).isRequired,
};

export default AppBoundary;
