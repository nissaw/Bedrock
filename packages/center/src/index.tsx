import styled from "styled-components";
import PropTypes from "prop-types";
import { mergeBreakpoints } from "@bedrock-layout/spacing-constants";

export interface CenterProps {
  maxWidth?: number;
}

const Center = styled.div<CenterProps>`
  --maxWidth: ${({ maxWidth, theme: { breakPoints } }) =>
    typeof maxWidth === "number"
      ? `${maxWidth}px`
      : mergeBreakpoints(breakPoints).medium + "px"};

  box-sizing: content-box;
  margin-left: auto;
  margin-right: auto;
  margin-inline-start: auto;
  margin-inline-end: auto;
  max-width: var(--maxWidth);
`;

Center.displayName = "Center";

Center.propTypes = {
  maxWidth: PropTypes.number,
};

export default Center;
