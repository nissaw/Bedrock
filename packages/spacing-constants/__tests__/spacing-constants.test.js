import * as constants from "../src";

describe("Spacing Constants", () => {
  test("constant snapshot", () => {
    expect(constants).toMatchSnapshot();
  });

  it("merges spacing", () => {
    const newSpacings = { xl: "8rem" };
    expect(constants.mergeSpacings(newSpacings)).toMatchSnapshot();
  });

  it("does not break if called without an object", () => {
    expect(constants.mergeSpacings()).toMatchSnapshot();
  });

  it("merges breakpoints", () => {
    const newBreakpoints = { xlarge: "1300px" };
    expect(constants.mergeBreakpoints(newBreakpoints)).toMatchSnapshot();
  });

  it("does not break if called with no object", () => {
    expect(constants.mergeBreakpoints()).toMatchSnapshot();
  });
});
