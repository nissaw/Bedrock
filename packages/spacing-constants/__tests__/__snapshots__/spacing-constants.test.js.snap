// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spacing Constants constant snapshot 1`] = `
Object {
  "breakPoints": Object {
    "large": 1199,
    "largeOnly": Array [
      1024,
      1199,
    ],
    "medium": 1023,
    "mediumOnly": Array [
      640,
      1023,
    ],
    "smallOnly": 639,
    "xlarge": 1439,
    "xlargeOnly": Array [
      1200,
      1439,
    ],
    "xxlarge": 1920,
  },
  "mergeBreakpoints": [Function],
  "mergeSpacings": [Function],
  "spacing": Object {
    "lg": "1rem",
    "lgXl": "1.5rem",
    "md": "0.5rem",
    "mdLg": "0.75rem",
    "none": "0px",
    "sm": "0.25rem",
    "xl": "2rem",
    "xlXXl": "3rem",
    "xs": "0.125rem",
    "xxl": "4rem",
  },
}
`;

exports[`Spacing Constants does not break if called with no object 1`] = `
Object {
  "large": 1199,
  "largeOnly": Array [
    1024,
    1199,
  ],
  "medium": 1023,
  "mediumOnly": Array [
    640,
    1023,
  ],
  "smallOnly": 639,
  "xlarge": 1439,
  "xlargeOnly": Array [
    1200,
    1439,
  ],
  "xxlarge": 1920,
}
`;

exports[`Spacing Constants does not break if called without an object 1`] = `
Object {
  "lg": "1rem",
  "lgXl": "1.5rem",
  "md": "0.5rem",
  "mdLg": "0.75rem",
  "none": "0px",
  "sm": "0.25rem",
  "xl": "2rem",
  "xlXXl": "3rem",
  "xs": "0.125rem",
  "xxl": "4rem",
}
`;

exports[`Spacing Constants merges breakpoints 1`] = `
Object {
  "large": 1199,
  "largeOnly": Array [
    1024,
    1199,
  ],
  "medium": 1023,
  "mediumOnly": Array [
    640,
    1023,
  ],
  "smallOnly": 639,
  "xlarge": "1300px",
  "xlargeOnly": Array [
    1200,
    1439,
  ],
  "xxlarge": 1920,
}
`;

exports[`Spacing Constants merges spacing 1`] = `
Object {
  "lg": "1rem",
  "lgXl": "1.5rem",
  "md": "0.5rem",
  "mdLg": "0.75rem",
  "none": "0px",
  "sm": "0.25rem",
  "xl": "8rem",
  "xlXXl": "3rem",
  "xs": "0.125rem",
  "xxl": "4rem",
}
`;
